ckanext/harvest/logic/action/update.py

```diff
@@ -530,6 +535,22 @@ def harvest_jobs_run(context, data_dict):
         context, {'source_id': source_id, 'status': u'Running'})
     if len(jobs):
         for job in jobs:
+            if timeout:
+                created = datetime.datetime.strptime(job['created'], '%Y-%m-%d %H:%M:%S.%f')
+                if datetime.datetime.now() - created > datetime.timedelta(minutes=int(timeout)):
+                    msg = 'Job timeout: %s is taking longer than %s minutes' % (job['id'], timeout)
+                    log.error(msg)
+
+                    job_obj = HarvestJob.get(job['id'])
+                    job_obj.status = u'Finished'
+                    job_obj.save()
```

Comment:
Should you also perhaps be setting the `finished` timestamp on this `job_obj` here? Seems to be the general pattern.

Also, correct me but setting this status won't actually _stop_ a job that's already running will it? So you'll still end up with a bunch of zombies running on the workers?

Reply:
I'll set the finished timestamp, although technically it didn't finish as it times out.

I believe that there's only 1 gather and 1 fetch worker running at any time. So we don't need to worry too much about zombie processes. Could add something in to check if the job is not set to Running during the fetch process if thats not already the case.

Reply:
> I'll set the finished timestamp, although technically it didn't finish as it times out.

Well that doesn't stop us setting its status to `Finished` though ;) - basically I'm just saying it's better not to put this job into a state that the system doesn't expect it to be in.

If there's only one worker though, that means that a zombie will block the pipeline entirely, no?

Reply:
I don't think a zombie will block the pipeline, what happens is that for some reason there are still harvest objects which are not completed, these depend on how the extensions extend the harvest logic and if it's not handled properly on the extension then the harvest object setting might not get updated so there are things still lying around on the queue. Which means that it can't set the job to finished. Could investigate issues on the extensions but I think it's good to have this timeout in place so that errors can bubble up to the user.